### PR TITLE
[feature](nereids) support timestampdiff function

### PR DESCRIPTION
**Pull request by englefly**
# Proposed changes
complete timeStampDiff
supported timeunit:
 - YEAR 
 - MONTH
 - WEEK
 - DAY
 - HOUR
 - MINUTE
 - SECOND

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.61 seconds
 load time: 492 seconds
 storage size: 17122795975 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221129053808_clickbench_pr_54780.html

**Comment by englefly** [comment]
@morrySnow PTAL

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4 [hunk 1]
```diff
@@ -257,8 +257,15 @@ valueExpression
...
 
+datetimeUnit
+    : YEAR | MONTH
+    | WEEK | DAY
+    | HOUR | MINUTE | SECOND
+    ;
+
 primaryExpression
-    : CASE whenClause+ (ELSE elseExpression=expression)? END                                   #searchedCase
+    : name=(TIMESTAMPDIFF | DATEDIFF) LEFT_PAREN unit=datetimeUnit COMMA startTimestamp=valueExpression COMMA endTimestamp=valueExpression RIGHT_PAREN    #timestampdiff
```
Comment (morrySnow) [inline fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4:267]: wrap this line, such as:
```suggestion
    : name=(TIMESTAMPDIFF | DATEDIFF) LEFT_PAREN unit=datetimeUnit COMMA
       startTimestamp=valueExpression COMMA endTimestamp=valueExpression RIGHT_PAREN    #timestampdiff
```

File: fe/fe-core/src/main/java/org/apache/doris/analysis/ArithmeticExpr.java [hunk 2]
```diff
@@ -62,7 +62,6 @@ public enum Operator {
         BITXOR("^", "bitxor", OperatorPosition.BINARY_INFIX, TExprOpcode.BITXOR),
         BITNOT("~", "bitnot", OperatorPosition.UNARY_PREFIX, TExprOpcode.BITNOT),
         FACTORIAL("!", "factorial", OperatorPosition.UNARY_POSTFIX, TExprOpcode.FACTORIAL);
-
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/ArithmeticExpr.java:65]: add this line back

File: fe/fe-core/src/main/java/org/apache/doris/analysis/FunctionCallExpr.java [hunk 3]
```diff
@@ -363,6 +363,16 @@ private String paramsToSql() {
...
+        if (fnName.getFunction().equalsIgnoreCase("years_diff")
+                || fnName.getFunction().equalsIgnoreCase("months_diff")
+                || fnName.getFunction().equalsIgnoreCase("days_diff")
+                || fnName.getFunction().equalsIgnoreCase("hours_diff")
+                || fnName.getFunction().equalsIgnoreCase("minutes_diff")
+                || fnName.getFunction().equalsIgnoreCase("seconds_diff")) {
+            sb.append(children.get(1).toSql()).append(", ");
+            sb.append(children.get(0).toSql()).append(")");
+            return sb.toString();
+        }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/FunctionCallExpr.java:375]: add comment to explain this is use for nereids and should not be removed

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 4]
```diff
@@ -613,7 +643,8 @@ public UnboundFunction visitFunctionCall(DorisParser.FunctionCallContext ctx) {
             //      the function information is obtained by parsing the catalog. This method is more scalable.
             String functionName = ctx.identifier().getText();
             boolean isDistinct = ctx.DISTINCT() != null;
-            List<Expression> params = visit(ctx.expression(), Expression.class);
+            List<ExpressionContext> expressionContexts = ctx.expression();
+            List<Expression> params = visit(expressionContexts, Expression.class);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:647]: why add a local var?

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4 (hunk 1 above)
Comment (englefly) [reply fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4:267]: done

File: fe/fe-core/src/main/java/org/apache/doris/analysis/ArithmeticExpr.java (hunk 2 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/ArithmeticExpr.java:65]: done

File: fe/fe-core/src/main/java/org/apache/doris/analysis/FunctionCallExpr.java (hunk 3 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/FunctionCallExpr.java:375]: done

Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/FunctionCallExpr.java:375]: I already add regression case for these functions. If someone remove the above code, the regression cases are failed.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java (hunk 4 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:647]: removed local var

**Review by morrySnow — APPROVED** [review]

**Pull request merged**